File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline
t = int(input())
for i in range(t):
    string = list(input())
    count = 0
    for j in range(len(string)):
        if string[j] == '(':
            count += 1
        elif string[j] == ')':
            count -= 1
        #^^... .j랑 i 구분 잘 하기.. 
        if count < 0:
            #만약 ))가 두개면 제대로 안 닫힌 거
            print('NO')
            break
    if count > 0:
        print('NO')
    elif count == 0:
        #else로 했는데 count < 0인 경우를 까먹음 ㅜ
        print('YES')

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            String str = br.readLine();
            Boolean bool = true; // 올바른 괄호인가? 확인용 
            int count = 0;
            
            for (char c : str.toCharArray()) {
                if (c == '(') {
                    // s == "(" 는 주소값을 비교함. char은 문자 비고
                    count++;
                } else {
                    count--;
                    if (count < 0) {
                        bool = false;
                        break;
                    }
                }
            }

            if (bool && count == 0) {
                // bool이 참이고, count가 0개면 = 개수가 같으면
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }

        }

    }
}